EssentialApp/EssentialAppTests/CommentsUIIntegrationTests.swift

```
@@ -0,0 +1,262 @@
+
+		sut.simulateUserInitiatedReload()
+		XCTAssertTrue(sut.isShowingLoadingIndicator, "Expected to show the loading indicator when the user initiates a reload")
+
+		loader.completeLoadingWithError(at: 1)
+		XCTAssertFalse(sut.isShowingLoadingIndicator, "Expected not to show the loading indicator after loading completed with an error")
+	}
+	
+	func test_loadCommentsCompletion_rendersSuccessfullyLoadedComments() {
+		let date = Date()
+		let comment0 = makeComment(id: UUID(), message: "message0", date: (date: date.adding(days: -1), string: "1 day ago"), author: "author0")
+		let comment1 = makeComment(id: UUID(), message: "message1", date: (date: date.adding(days: -3), string: "3 days ago"), author: "author1")
+		let comment2 =  makeComment(id: UUID(), message: "message2", date: (date: date.adding(days: -31), string: "1 month ago"), author: "author2")
+		let comment3 = makeComment(id: UUID(), message: "message3", date: (date: date.adding(days: -366), string: "1 year ago"), author: "author3")
+		let (sut, loader) = makeSUT(currentDate: { date }, locale: .init(identifier: "en_US_POSIX"))
+
+		sut.loadViewIfNeeded()
+		assertThat(sut, isRendering: [])
+
+		loader.completeLoading(with: [comment0.model], at: 0)
+		assertThat(sut, isRendering: [comment0.expected])
+
+		sut.simulateUserInitiatedReload()
+		loader.completeLoading(with: [comment0.model, comment1.model, comment2.model, comment3.model], at: 1)
+		assertThat(sut, isRendering: [comment0.expected, comment1.expected, comment2.expected, comment3.expected])
+	}
+	
+	func test_loadCommentsCompletion_rendersErrorMessageOnLoaderFailureUntilNextReload() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+
+		loader.completeLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("COMMENTS_VIEW_CONNECTION_ERROR"))
+
+		sut.simulateUserInitiatedReload()
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
```

- Missing test/assertion to check that tapping on the error view will hide the error.

- Missing test/assertion to check that an error doesn't alter the current state of the comments view.

Check the Feed UI integration tests as a guideline.

---

done